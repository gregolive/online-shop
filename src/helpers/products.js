import uniqid from 'uniqid';

import pearlImg from '../img/Acaia-Pearl-S-black.jpg';
import tampImg from '../img/forcetamp585.jpg';
import basketImg from '../img/ims18g-ridgeless.jpg';
import knockboxImg from '../img/maple-knock-set.jpg';
import burrImg from '../img/ssp-multi-purpose.jpg';
import matImg from '../img/tamping-mat-corner.jpg';

class Product {
  constructor (name, vendor, price, stock, img) {
    this.id = uniqid();
    this.name = name;
    this.vendor = vendor;
    this.price = this.formatPrice(price);
    this.stock = stock;
    this.img = img;
  }

  inStock() { return (this.stock > 0) ? true : false };

  formatPrice = (value) => {
    const dollars = Math.trunc(value / 100);
    let cents = value % 100;
    if (cents < 10) { cents = '0' + cents; }
    return `$${dollars}.${cents}`
  };
};

const products = [
  new Product('Pearl Model S Coffee Scale - Black', 'Acaia', 19500, 5, pearlImg),
  new Product('The Force Tamp 58.5mm', 'Espresso Parts', 19900, 2, tampImg),
  new Product('Precision Portafilter Basket - Double 18g - Ridgeless', 'IMS', 3058, 4, basketImg),
  new Product('Maplewood Knockbox Set', 'IMS', 6396, 10, knockboxImg),
  new Product('Multi-Purpose Grinding Red Speed Coated 64mm Flat Burrs', 'SSP', 18500, 1, burrImg),
  new Product('Tamping Mat with Corner Edge', 'Barista Basics', 2800, 0, matImg),
];

export default products;

const findProduct = (productName) => products.filter((p) => p.name === productName)[0];

export { findProduct };
