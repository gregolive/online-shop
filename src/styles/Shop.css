.ProductGrid {
  display: grid;
  grid-template-columns: repeat(auto-fit, minmax(240px, 1fr));
  gap: 2rem;
}

.Product {
  color: var(--text);
  display: grid;
  grid-template-rows: auto 1fr auto;
  gap: 1rem;
}

.ShopImg {
  width: 100%;
}

.ShopVendor {
  color: var(--highlight);
}

.ProductName > a {
  transition: all 0.2s ease;
}

.ProductName > a:hover {
  color: var(--highlight);
}

.CartButton {
  color: var(--text);
  font-size: 1rem;
}

.CartButton:hover {
  background: var(--secondary);
}
