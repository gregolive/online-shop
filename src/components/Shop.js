import '../styles/Shop.css';
import { Link } from "react-router-dom";
import products from '../helpers/products';

const Shop = ({ updateCart }) => {
  return (
    <section className="Shop">
      <h2 className='ShopHeading'>Shop Espresso Tools</h2>
      <div className='ProductGrid'>
        {products.map((product) =>
          <article key={product.id} className='ShopProduct'>
            <Link to={product.name}>
              <img src={product.img} alt={`${product.vendor} - ${product.name}`} className='ShopImg' />
            </Link>

            <div>
              <small className='ShopVendor'>{product.vendor}</small>
              <h3 className='ProductName'>
                <Link to={product.name}>{product.name}</Link>
              </h3>
              <strong className='ShopPrice'>{product.price}</strong>
            </div>

            {(product.inStock()) ? 
              <button type='button' className='Button PrimaryButton' onClick={() => updateCart(product, 1)}>Add to cart</button>
              : <button type='button' className='Button PrimaryButton' disabled>Add to cart</button>
            }
          </article>
        )}
      </div>
    </section>
  );
};

export default Shop;
