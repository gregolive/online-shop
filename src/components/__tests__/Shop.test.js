import { render, screen } from '@testing-library/react';
import { BrowserRouter } from "react-router-dom";
import Shop from '../Shop';
import products from '../../helpers/products';

test('renders shop heading', () => {
  render(<BrowserRouter><Shop /></BrowserRouter>);
  expect(screen.getAllByRole('heading')[0].textContent).toMatch(/Shop Espresso Tools/);
});

test('renders products', () => {
  render(<BrowserRouter><Shop /></BrowserRouter>);
  expect(screen.getAllByRole('article').length).toBeGreaterThan(0);
});

const testProduct = products[0];

test('renders product image', () => {
  render(<BrowserRouter><Shop /></BrowserRouter>);
  expect(screen.getByAltText(`${testProduct.vendor} - ${testProduct.name}`)).toBeInTheDocument();
});

test('renders product vendor', () => {
  render(<BrowserRouter><Shop /></BrowserRouter>);
  expect(screen.getAllByText(testProduct.vendor)[0]).toBeInTheDocument();
});

test('renders product name', () => {
  render(<BrowserRouter><Shop /></BrowserRouter>);
  expect(screen.getByText(testProduct.name)).toBeInTheDocument();
});

test('renders product price', () => {
  render(<BrowserRouter><Shop /></BrowserRouter>);
  expect(screen.getAllByText(testProduct.price)[0]).toBeInTheDocument();
});

test('renders add to cart button', () => {
  render(<BrowserRouter><Shop /></BrowserRouter>);
  expect(screen.getAllByRole('button')[0].textContent).toMatch(/Add to cart/);
});

it('renders consistently', () => {
  const { container } = render(<BrowserRouter><Shop /></BrowserRouter>);
  expect(container).toMatchSnapshot();
});
