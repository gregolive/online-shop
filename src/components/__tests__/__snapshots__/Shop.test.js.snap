// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders consistently 1`] = `
<div>
  <section
    class="Shop"
  >
    <h2
      class="ShopHeading"
    >
      Shop Espresso Tools
    </h2>
    <div
      class="ProductGrid"
    >
      <article
        class="Product"
      >
        <a
          href="/Pearl Model S Coffee Scale - Black"
        >
          <img
            alt="Acaia - Pearl Model S Coffee Scale - Black"
            class="ProductImg"
            src="Acaia-Pearl-S-black.jpg"
          />
        </a>
        <div
          class="ProductOverview"
        >
          <small
            class="Vendor"
          >
            Acaia
          </small>
          <h3
            class="ProductName"
          >
            <a
              href="/Pearl Model S Coffee Scale - Black"
            >
              Pearl Model S Coffee Scale - Black
            </a>
          </h3>
          <strong
            class="Price"
          >
            $195.00
          </strong>
        </div>
        <button
          class="Button CartButton"
        >
          Add to cart
        </button>
      </article>
      <article
        class="Product"
      >
        <a
          href="/The Force Tamp 58.5mm"
        >
          <img
            alt="Espresso Parts - The Force Tamp 58.5mm"
            class="ProductImg"
            src="forcetamp585.jpg"
          />
        </a>
        <div
          class="ProductOverview"
        >
          <small
            class="Vendor"
          >
            Espresso Parts
          </small>
          <h3
            class="ProductName"
          >
            <a
              href="/The Force Tamp 58.5mm"
            >
              The Force Tamp 58.5mm
            </a>
          </h3>
          <strong
            class="Price"
          >
            $199.00
          </strong>
        </div>
        <button
          class="Button CartButton"
        >
          Add to cart
        </button>
      </article>
      <article
        class="Product"
      >
        <a
          href="/Precision Portafilter Basket - Double 18/20 gr - Ridgeless"
        >
          <img
            alt="IMS - Precision Portafilter Basket - Double 18/20 gr - Ridgeless"
            class="ProductImg"
            src="ims18g-ridgeless.jpg"
          />
        </a>
        <div
          class="ProductOverview"
        >
          <small
            class="Vendor"
          >
            IMS
          </small>
          <h3
            class="ProductName"
          >
            <a
              href="/Precision Portafilter Basket - Double 18/20 gr - Ridgeless"
            >
              Precision Portafilter Basket - Double 18/20 gr - Ridgeless
            </a>
          </h3>
          <strong
            class="Price"
          >
            $30.58
          </strong>
        </div>
        <button
          class="Button CartButton"
        >
          Add to cart
        </button>
      </article>
      <article
        class="Product"
      >
        <a
          href="/Maplewood Knockbox Set"
        >
          <img
            alt="IMS - Maplewood Knockbox Set"
            class="ProductImg"
            src="maple-knock-set.jpg"
          />
        </a>
        <div
          class="ProductOverview"
        >
          <small
            class="Vendor"
          >
            IMS
          </small>
          <h3
            class="ProductName"
          >
            <a
              href="/Maplewood Knockbox Set"
            >
              Maplewood Knockbox Set
            </a>
          </h3>
          <strong
            class="Price"
          >
            $63.96
          </strong>
        </div>
        <button
          class="Button CartButton"
        >
          Add to cart
        </button>
      </article>
      <article
        class="Product"
      >
        <a
          href="/Multi-Purpose Grinding Red Speed Coated 64mm Flat Burrs"
        >
          <img
            alt="SSP - Multi-Purpose Grinding Red Speed Coated 64mm Flat Burrs"
            class="ProductImg"
            src="ssp-multi-purpose.jpg"
          />
        </a>
        <div
          class="ProductOverview"
        >
          <small
            class="Vendor"
          >
            SSP
          </small>
          <h3
            class="ProductName"
          >
            <a
              href="/Multi-Purpose Grinding Red Speed Coated 64mm Flat Burrs"
            >
              Multi-Purpose Grinding Red Speed Coated 64mm Flat Burrs
            </a>
          </h3>
          <strong
            class="Price"
          >
            $185.00
          </strong>
        </div>
        <button
          class="Button CartButton"
        >
          Add to cart
        </button>
      </article>
      <article
        class="Product"
      >
        <a
          href="/Tamping Mat with Corner Edge"
        >
          <img
            alt="Barista Basics - Tamping Mat with Corner Edge"
            class="ProductImg"
            src="tamping-mat-corner.jpg"
          />
        </a>
        <div
          class="ProductOverview"
        >
          <small
            class="Vendor"
          >
            Barista Basics
          </small>
          <h3
            class="ProductName"
          >
            <a
              href="/Tamping Mat with Corner Edge"
            >
              Tamping Mat with Corner Edge
            </a>
          </h3>
          <strong
            class="Price"
          >
            $28.00
          </strong>
        </div>
        <button
          class="Button CartButton"
        >
          Add to cart
        </button>
      </article>
    </div>
  </section>
</div>
`;
